kernel/arch/dreamcast/sound/snd_stream.c

```diff
@@ -675,16 +634,128 @@ static inline void dma_chain(void *data) {
     }
 }
 
+static void snd_stream_transfer(strchan_t *stream, void *first_buf,
+                                uint32_t offset, size_t size) {
+    int rs;
+
+    dcache_purge_range((uintptr_t)first_buf, size);
+    stream->mutex_thd = thd_current;
+
+    if(stream->channels == 2) {
+        dcache_purge_range((uintptr_t)sep_buffer[1], size);
+        stream->dma_dest = stream->spu_ram_sch[1] + offset;
+        stream->dma_length = size;
+    }
+
+    do {
+        rs = spu_dma_transfer(first_buf,
+            stream->spu_ram_sch[0] + offset,
+            size,
+            0,
+            (stream->channels == 1 ? dma_done : dma_chain),
+            (void *)stream);
+
+        if(rs == 0) {
+            break;
+        }
+        thd_pass();
+    } while(1);
+}
+
+static size_t snd_stream_fill(snd_stream_hnd_t hnd, uint32_t offset, size_t size) {
+    strchan_t *stream = &streams[hnd];
+    const int chans = stream->channels;
+    const uintptr_t left = stream->spu_ram_sch[0] + offset;
+    const uintptr_t right = stream->spu_ram_sch[1] + offset;
+    const int needed_bytes = size * chans;
+    int got_bytes = 0;
+    void *data = NULL;
+
+    if(stream->req_data) {
+        got_bytes = stream->req_data(hnd,
+            (left | SPU_RAM_UNCACHED_BASE),
+            (chans == 2 ? (right | SPU_RAM_UNCACHED_BASE) : 0),
+            needed_bytes);
+    }
+    if(got_bytes > 0) {
+        return got_bytes;
+    }
+    if(stream->get_data) {
+        data = stream->get_data(hnd, needed_bytes, &got_bytes);
+    }
+
+    if(data == NULL || got_bytes == 0) {
+        /* sep_buffer isn't allocated if all streams are mono */
+        if(sep_buffer[0] == NULL) {
```

Comment:
Is `sep_buffer[0] == NULL` the same as checking if `chans == 1`?  If so, then can we drop the `if(chans == 2)` check on line 695?

Reply:
I just realized that I didn't fix the comment in the code! Thanks for paying attention, an addition needs to be made.